Use a HashSet<BlockRef> for a few invariants like in C YJIT

XrXr opened the pull request:
It's important that we are able to quickly remove a block from the
Invariants struct and keeping the BlockRefs in a Vec meant we need
a linear search for the removal.

Implement by-identity Hash and Eq for BlockRef and use HashSet for
`single_ractor`, `basic_operators`, and `global_constant_state`.
This is like in C YJIT, where we use `st_table` as a hash set for
the blocks.

maximecb commented on yjit/src/core.rs:
```diff
@@ -703,7 +742,7 @@ impl Block {
         // Wrap the block in a reference counted refcell
         // so that the block ownership can be shared
-        Rc::new(RefCell::new(block))
+        BlockRef(Rc::new(RefCell::new(block)))
```
I think it would be nice to have a `BlockRef::new` constructor here. Will make an eventual transition to Box easier too.

maximecb commented on yjit/src/invariants.rs:
```diff
@@ -128,13 +124,24 @@ pub fn assume_single_ractor_mode(jit: &mut JITState, ocb: &mut OutlinedCb) -> bo
+    for block in &blocks {
         invalidate_block_version(block);
         incr_counter!(invalidate_bop_redefined);
     }
```
In this specific case I think it would be really nice to keep the functionality of being able to index by BOP.

maximecb replied:
Sidenote that @kddnewton  is coming back tomorrow so possibly he could complete the bop invalidation code if you'd just rather keep going with the rest of `invalidate_block_version(), or you two could pair on it.

XrXr replied:
We also need to be able to index by block, though, invalidation needs it. If we want to switch this setup we'd need to keep some more data on the block so invalidation can remove the block from the table without needing to go through every block in here.

Freeing dead iseqs also wants quick lookup/removal.

maximecb replied:
Right, need another table that maps blocks to which assumptions they make.

maximecb replied:
I don't want to slow you down, so I think I will ask Kevin to implement the bop-based invalidation. Can disregard this comment and just take care of the other one.

maximecb reviewed (commented):
Looks pretty good with two comments.

kddnewton committed:
Allow quickly invalidating a BOP or a block

Tracks blocks to BOPs and BOPs to blocks to allow quickly invalidating either one.

XrXr committed:
Use a HashSet<BlockRef> for a few invariants like in C YJIT

It's important that we are able to quickly remove a block from the
Invariants struct and keeping the BlockRefs in a Vec meant we need
a linear search for the removal.

Implement by-identity Hash and Eq for BlockRef and use HashSet for
`single_ractor`, `basic_operators`, and `global_constant_state`.
This is like in C YJIT, where we use `st_table` as a hash set for
the blocks.

XrXr committed:
Use remove, no functors so plain text is enough to see what's going on

XrXr committed:
Allow quickly invalidating a BOP or a block
